### PR TITLE
Streak check improvements

### LINKED ISSUE #118: Streak check improvements
Streak check currently only looking at simple straight runs of the same value.

Think about other options (repeated days, same value at same hour for a number of days).

Be careful with wind speeds (calms) and wind directions (low resolution) on how this is best managed.

### PR DESCRIPTION
**Pull request by rjhd2**
Closes #118

### CODE REVIEW

File: inter_checks.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ def run_checks(restart_id:str = "", end_id:str = "", diagnostics:bool = False, p
 
         try:
             target_station, target_station_df = io.read_station(os.path.join(
-                setup.SUBDAILY_PROC_DIR, "{target_station_id:11s}.qff{setup.OUT_COMPRESSION}"),
+                setup.SUBDAILY_PROC_DIR, f"{target_station_id:11s}.qff{setup.OUT_COMPRESSION}"),
```
Comment (rjhd2) [inline inter_checks.py:126]: bugfix for missing `f` in f-string format

File: qc_tests/humidity.py [hunk 2]
```diff
@@ -40,11 +40,11 @@ def get_repeating_dpd_threshold(temperatures: utils.Meteorological_Variable,
 
     # only process further if there are enough locations
     if len(locs) > 1:
-        repeated_string_lengths, _, _ = utils.prepare_data_repeating_string(locs, diff=1, plots=plots, diagnostics=diagnostics)
+        repeated_streak_lengths, _, _ = utils.prepare_data_repeating_streak(locs, diff=1, plots=plots, diagnostics=diagnostics)
```
Comment (rjhd2) [inline qc_tests/humidity.py:43]: standardised naming so that calling these _streaks_ not _strings_ (which is probably safer giving typing)

**Comment by rjhd2** [comment]
Perhaps to reduce both duplication and complexity, the repeating streak and excess streak routines could be done as the repeating day routine, i.e. one routine, but with a calculate/set switch.  That ensures the data preparation is the same for both steps perhaps at the cost of a more complex routine.

File: plot_map_of_flagging_rates.py [hunk 3]
```diff
@@ -13,12 +13,17 @@
...
+                                  "S", "U", "V", "W", "d", "o", "x", "y",],
+                  "dew_point_temperature": ['All', 'C', 'D', 'E', 'F', 'H', 'K', 'L', 'N', 'S',
+                                            'T', 'V', 'W', 'd', 'h', 'n', 'o', 'x', 'y'],
+                  "sea_level_pressure" : ['All', 'D', 'E', 'F', 'H', 'K', 'L', 'N', 'S',
+                                          'T', 'V', 'W', 'd', 'o', 'p', 'x', 'y'],
+                  "station_level_pressure" : ['All', 'D', 'E', 'F', 'H', 'K', 'L', 'N',
+                                              'S', 'T', 'V', 'd', 'o', 'p', 'x', 'y'],
+                  "wind_speed" : ['All', 'E', 'H', 'K', 'L', 'N', 'S', 'T', 'V',
+                                  'W', 'o', 'w', 'x', 'y'],
+                  "wind_direction" : ['All', 'E', 'H', 'K', 'L', 'w', 'x', 'y']}
```
Comment (rjhd2) [inline plot_map_of_flagging_rates.py:26]: Reordering for ease of maintenance (done when adding new flags for streak tests)

File: qc_tests/streaks.py [hunk 4]
```diff
@@ -4,28 +4,33 @@
...
 import copy
 import numpy as np
+import datetime as dt
 import logging
 logger = logging.getLogger(__name__)
+from typing import Optional
 
 import qc_utils as utils
 
+MIN_STREAK_LENGTH_FOR_EXCESS_FREQUENCY = 10
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:23]: Are these consecutive hours or consecutive obs? So obs could be 8 hourly therefor on 2 obs may be needed to identify a 10 hour streak compared to 10 obs for hourly data..

**Review by Kate-Willett — COMMENTED** [review]
Hopefully this review has worked. I've just reviewed the amended bits and a few bits inbetween that I looked at to try and follow the code.

File: qc_tests/streaks.py [hunk 5]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
 
+    if threshold == utils.MDI:
+        # No threshold obtainable, no need to continue the test
+        return
+    
     # only process further if there is enough data
-    if len(this_var.data.compressed()) > 1:
-        repeated_string_lengths, grouped_diffs, strings = utils.prepare_data_repeating_string(this_var.data.compressed(), diff=0, plots=plots, diagnostics=diagnostics)
+    if len(this_var.data.compressed()) < 2:
+        # Escape if insufficient data
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:286]: Think I've already commented here that its not obvious what 'enough' data is. I'm not familiar with .compressed() to know why you would change from > 1 to < 2. So more commenting would be helpful here?

File: qc_tests/streaks.py [hunk 6]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
-        # above threshold
-        bad, = np.where(repeated_string_lengths >= threshold["Straight"])
+    repeated_streak_lengths, grouped_diffs, streaks = utils.prepare_data_repeating_streak(this_var.data.compressed(), diff=0, plots=plots, diagnostics=diagnostics)
 
-        # flag identified strings
-        for string in bad:
-            start = int(np.sum(grouped_diffs[:strings[string], 1]))
-            end = start + int(grouped_diffs[strings[string], 1]) + 1
+    # above threshold
+    bad, = np.where(repeated_streak_lengths >= threshold)
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:292]: Use of np.where could be np.nonzero to protect against numpy 2.0 changes to np.where?

File: qc_tests/streaks.py [hunk 7]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
 
 #************************************************************************
-def excess_repeating_value():
+def excess_repeating_value(obs_var: utils.Meteorological_Variable, times: np.array,
+                    config_dict: dict, 
+                    wind_speed: Optional[utils.Meteorological_Variable] = None,
+                    plots: bool = False, diagnostics: bool = False) -> None:
+    """
+    Flag years where more than expected fraction of data occurs in streaks,
+      but none/not many are long enough in themselves to trigger the repeating_value check
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:321]: This is still a bit confusing. 'excess_repeating_value' sounds like a test for overly frequent values but here you talk about short streaks repeating. So is this different from individual values that get erroneously repeated throughout the record?

File: qc_tests/streaks.py [hunk 8]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
+        # Not looking for distribution of streak lengths,
+        #  but proportion of obs identified as in a streak in each calendar year.
+        (year_repeated_streak_lengths,
+        grouped_diffs,
+        streaks) = utils.prepare_data_repeating_streak(this_var.data[locs].compressed(),
+                                                        diff=0, plots=plots,
+                                                        diagnostics=diagnostics)
+
+        long_enough_streaks, = np.where(year_repeated_streak_lengths >= MIN_STREAK_LENGTH_FOR_EXCESS_FREQUENCY)
+
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:379]: Use of np.where which changes when numpy 2.0 used?

File: qc_tests/streaks.py [hunk 9]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
+    flags = np.ma.array(["" for i in range(this_var.data.shape[0])],
+                        mask=this_var.data.mask)
+
+    # retrieve the threshold and store
+    try:
+        threshold = config_dict[f"STREAK-{this_var.name}"]["Excess"]
+    except KeyError:
+        # no threshold set
+        get_excess_streak_threshold(this_var, years, config_dict,
+                                    plots=plots, diagnostics=diagnostics)
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:351]: Check naming convention with excess repeated values or is it repeated streaks of values?

File: qc_tests/streaks.py [hunk 10]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
+
+    # sort into array of days
+    # check each day and compare to previous. HadISD used fixed threshold, but could do one pass to get, second to apply/?
+    all_lengths = []
+    streak_length = 0
+    previous_day_data = np.array([])
+    streak_locs = np.array([], dtype=(int))
+
+    for year in np.unique(station.years):
+        for month in np.unique(station.months):
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:434]: Does this only search within the year assigned in for loop above, and ditto for station.days below? Otherwise there I don't see how it will work logically through each day, month, year?

File: qc_tests/streaks.py [hunk 11]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
+            for day in np.unique(station.days):
+                try:
+                    # if Datetime doesn't throw an error, then valid date
+                    _ = dt.datetime(year, month, day)
+                except ValueError:
+                    # not a valid day (e.g. Leap years, short months etc)
+                    continue
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:441]: Why is a leap year not valid? What do you do with leap years?

File: qc_tests/streaks.py [hunk 12]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
+
+                this_day_locs, = np.where(np.logical_and.reduce((station.years == year,
+                                                        station.months == month,
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:444]: Clever! Check use of np.where if you move to numpy 2.0?

File: qc_tests/streaks.py [hunk 13]
```diff
@@ -93,102 +170,354 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
-def hourly_repeat():
+#def hourly_repeat():
 
     # repeat of given value at same hour of day for > N days
     # HadISD used fixed threshold.  Perhaps can dynamically calculate?
 
-    return # hourly_repeat
+    # sort in to 24hr expanded array
+    # for each hour
+    # check each day, HadISD used fixed threshold, but could do one pass to get, second to apply/?
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:511]: Should these comments be within a function?

File: qc_tests/streaks.py [hunk 14]
```diff
@@ -203,29 +532,50 @@ def rsc(station, var_list, config_dict, full=False, plots=False, diagnostics=Fal
...
+            wind_speed = copy.deepcopy(getattr(station, "wind_speed"))
+        else:
+            wind_speed = None
 
         # need to have at least two observations to get a streak
         if len(obs_var.data.compressed()) >= 2:
-
-            # repeating (straight) strings
             if full:
-                get_repeating_string_threshold(obs_var, config_dict, plots=plots, diagnostics=diagnostics)
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:212]: Similar terminology to streak testing. Is it using same function or could it be uniquely named?

File: qc_tests/streaks.py [hunk 15]
```diff
@@ -203,29 +532,50 @@ def rsc(station, var_list, config_dict, full=False, plots=False, diagnostics=Fal
...
+        if obs_var.name == "wind_direction":
+            wind_speed = copy.deepcopy(getattr(station, "wind_speed"))
+        else:
+            wind_speed = None
 
         # need to have at least two observations to get a streak
         if len(obs_var.data.compressed()) >= 2:
-
-            # repeating (straight) strings
             if full:
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:542]: What is 'full'? What does it mean?

File: qc_tests/streaks.py [hunk 16]
```diff
@@ -203,29 +532,50 @@ def rsc(station, var_list, config_dict, full=False, plots=False, diagnostics=Fal
...
+                # recalculating all thresholds
+                get_repeating_streak_threshold(obs_var, config_dict, wind_speed=wind_speed,
+                                               plots=plots, diagnostics=diagnostics)
+                get_excess_streak_threshold(obs_var, station.years, config_dict, wind_speed=wind_speed,
+                                            plots=plots, diagnostics=diagnostics)
+
+                # This is slow - as spinning through each day and then selecting
+                repeating_day(obs_var, station, config_dict, determine_threshold=True,
+                              plots=plots, diagnostics=diagnostics)
+
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:552]: Double check indentation here. I think its right but its hard to understand the logic a bit.

File: qc_tests/streaks.py [hunk 17]
```diff
@@ -4,28 +4,33 @@
 
    Checks for replication of 
      1. checks for consecutive repeating values
-     2. checks if one year has more repeating strings than expected [Not yet implemented]
+     2. checks if one year has more repeating streaks than expected
      3. checks for repeats at a given hour across a number of days [Not yet implemented]
-     4. checks for repeats for whole days - all 24 hourly values [Not yet implemented]
+     4. checks for repeats for whole days - all 24 hourly values
```
Comment (Kate-Willett) [inline qc_tests/streaks.py:9]: Double check this explanation tallies up with the QC tests/functions within. It would be easier to follow if you add the relevant function names to the docstrings here I think. I'm not entirely clear what tests are done here or where the humidity variables only have persistent 0 DPD tested or whether that is in addition?

File: qc_utils.py [hunk 18]
```diff
@@ -45,6 +45,8 @@
             "o" : "Odd Cluster",
             "p" : "Pressure",
             "w" : "Winds",
+            "x" : "Excess streak proportion",
+            "y" : "Repeated Day streaks"
             }
```
Comment (Kate-Willett) [inline qc_utils.py:50]: Is the 'humidity' one supersaturation or repeated streaks of 0 DPD?

File: qc_utils.py [hunk 19]
```diff
@@ -345,45 +350,66 @@ def gcv_calculate_binmax(indata: np.array, binmin: float, binwidth: float) -> fl
     return binmax  #gcv_calculate_binmax
     
 
-
-def gcv_central_section(full_hist: np.array) -> tuple[int, int]:
+#*********************************************
+def gcv_zeros_in_central_section(histogram: np.array, inner_n: int) -> tuple[int, int]:
     """
     A routine to determine if, for a distribution with multiple peaks, the central
     section is sufficiently big.
```
Comment (Kate-Willett) [inline qc_utils.py:357]: What meteorological issue is this trying to find? Something unusual about the distribution but can you give more info here in the docstring? Why do we believe this to be an issue?

File: qc_utils.py [hunk 20]
```diff
@@ -392,123 +418,139 @@ def get_critical_values(indata: np.array, binmin: float = 0, binwidth: float = 1
     :param float binwidth: bin width
     :param bool plots: do the plots
     :param bool diagnostics : do diagnostic outputs
-    :param str line_label: label for plotted histogram
+    :param str line_label: label for ploted histogram
```
Comment (Kate-Willett) [inline qc_utils.py:421]: Typo ploted?

File: qc_utils.py [hunk 21]
```diff
@@ -392,123 +418,139 @@ def get_critical_values(indata: np.array, binmin: float = 0, binwidth: float = 1
...
+    #  Avoids risk of secondary populations in the distribution affecting the fit
+    edges = full_edges[:10]
+    central_hist = full_hist[:10]
 
-                            # find first empty bin after that
-                            first_zero_bin, = np.argwhere(full_hist[fit_below_point1:] == 0)[0]
-                            threshold = binwidth * (binmin + fit_below_point1 + first_zero_bin)
+    # Remove zeros (turn into infs in log-space)
+    goods, = np.where(central_hist != 0)
+    hist = central_hist[goods]
```
Comment (Kate-Willett) [inline qc_utils.py:474]: CHeck use of np.where.

File: qc_utils.py [hunk 22]
```diff
@@ -947,14 +989,23 @@ def find_continent(country_code: str) -> str:
...
     :param np.array data: data to assess
-    :param int diff: difference to look for (0 in strings of data, 1 in strings of indices)
+    :param int diff: difference to look for (0 in streaks of data, 1 in streaks of indices)
     :param bool plots: turn on plots
     :param bool diagnostics: turn on diagnostic output
+
+    :returns: tuple(array, array, array)
+
+    array of the streak lengths in temporal order
+    array of the grouped differences (difference, count)
```
Comment (Kate-Willett) [inline qc_utils.py:1007]: Not sure what grouped differences are?

**Review by Kate-Willett — COMMENTED** [review]
I've gone through all changes in main scripts but not got to tests yet.

Generally this seems good. I've found it hard to follow logic in many places but that's partly because I'm not looking at full scripts and also because its a beast! 

My comments are short only because I'm trying to be time efficient - please do not read as grumpy in any way!

Some lines are a little on long side but I can see its best to focus on what the code does and comments rather than perfect code prettiness.

File: qc_tests/humidity.py [hunk 23]
```diff
@@ -222,15 +222,15 @@ def dew_point_depression_streak(times: np.array,
 
     # only process further if there are enough locations
     if len(locs) > 1:
-        repeated_string_lengths, grouped_diffs, strings = utils.prepare_data_repeating_string(locs, diff=1, plots=plots, diagnostics=diagnostics)
+        repeated_streak_lengths, grouped_diffs, streaks = utils.prepare_data_repeating_streak(locs, diff=1, plots=plots, diagnostics=diagnostics)
 
         # above threshold
-        bad, = np.where(repeated_string_lengths >= threshold)
+        bad, = np.where(repeated_streak_lengths >= threshold)
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply qc_tests/humidity.py:229]: Yes, I do have an environment, but it's a fair point.  I'll address these here as it's just a swap out as the functionality is the same (though, to my mind, less transparent as to what the code is doing).  I'll make a new issue to address all `np.where`s in the repo.

File: qc_utils.py (hunk 18 above)
Comment (rjhd2) [reply qc_utils.py:50]: Both are captured by the "h" flag rather than giving these two issues separate flag characters.  Happy to discuss whether that is something to address.

File: qc_utils.py (hunk 19 above)
Comment (rjhd2) [reply qc_utils.py:357]: Hopefully I've improved the docstring.  This is a helper routine for the get_critical_values() function to count how many zero-valued bins there are in the `inner_n` bins, as if many are zero, then the x^-1 line fitting didn't work well on some occaisions.  However, your comment did highlight that the typing didn't reflect the current format of this function, so I've updated that.

File: qc_utils.py (hunk 22 above)
Comment (rjhd2) [reply qc_utils.py:1007]: In this case it pairs of the difference between values (e.g. 0 for a repeating string) and the number of times this occurs before the difference changes.

File: qc_tests/streaks.py (hunk 5 above)
Comment (rjhd2) [reply qc_tests/streaks.py]: I changed the logic - rather than lots of nested if-loops, I've changed where I could so that the subroutine returns if something isn't met.  This reduces the indenting.  So rather than "continue processing if data array has > 1 value", it's now "exit if data array has < 2 values".

I've done "<2" so that both length 0 and length 1 arrays are captured

File: qc_tests/streaks.py (hunk 6 above)
Comment (rjhd2) [reply qc_tests/streaks.py]: Good call - see response elsewhere for this PR - have changed here, and raised an issue to fix throughout code.

File: qc_tests/streaks.py (hunk 7 above)
Comment (rjhd2) [reply qc_tests/streaks.py:325]: I've called this "annual string expectance" in HadISD - i.e. more observations in short strings than you'd expect (presuming most years are "good" data).  Would that be a better name?  I agree, there could be some confusion with the "frequent_values" check - though in that case there's no requirement for values to be in a chain of repeating ones.  Here, there has to be some repetition aspect - so lots of short strings, and more than expected (and excess number) - but none themselves long enough to set off the other test.

File: qc_tests/streaks.py (hunk 9 above)
Comment (rjhd2) [reply qc_tests/streaks.py:355]: I'll amend this naming once the test name is finalised.

File: qc_tests/streaks.py (hunk 8 above)
Comment (rjhd2) [reply qc_tests/streaks.py:383]: Done

File: qc_tests/streaks.py (hunk 12 above)
Comment (rjhd2) [reply qc_tests/streaks.py:461]: Done - np.nonzero(x in array) is the exact equivalent of np.where(x in array).  np.where() does other things if you e.g. pass in extra arguments, which I've rarely used, and will ensure that I replace each instance manually to check.

**Review by pwhybra — COMMENTED** [review]
```diff
# Lines 
+1234
-245
```


There are quite a lot of changes in this PR. So a little challenging for me as the first time looking at this repo. 

As such, __apologies__ for the superficial comments to start with as I work through it.




> [!NOTE]
> 
>The current environment.yml file in this PR seemed to error when trying to build.
>I needed a slightly different environment and configuration.txt file to get going.
>


@rjhd2  shared another version (which I have temporarily added [here](https://gist.github.com/pwhybra/b6c5925e477818553fb4abe21b420435)).

After using this environment, I was able to run the steaks checks and all worked well.

Note that this environment does not contain sphinx so docs cannot be built, though they are not updated for this PR.



I would recommend a small PR before merging this one to sync and create a coherent 
environment, as I think this is perhaps more critical than the PR to be able to review it.

I think you may have started this (?), though happy to help do that as shouldn't take long!)


### Reviewing/Running tests

Once setup all test run and pass! :)

### Structure of repo and testing coverage

Looking at test coverage via:
```terminal
coverage run -m pytest
coverage html
```

Currently, all of the tests are included in the coverage report by default, which isn't ideal. A small change to the structure of the repo could fix this, but you can also get around it via:

```terminal
coverage run --omit 'tests/*' -m pytest
```

`streaks.py` looks well covered by tests, and I have started to look through them with a few comments here and there.

There are tests that seem to _test_ the generation of the test data, which is quite meta!

There are a few it would be good to discuss with you, as I didn't quite follow what was going on.


It is nice to see a lot of testing! There is a lot more mocking than I normally do, so good to learn more and remind myself about that.

### Small refactors to reduce nesting

I suggested one example, but there are quite a few cases where you could restructure the logic in a function slightly to remove nesting, which I think makes things easier to follow.

We could chat about a few of those if interested, but that is more about style and readability than correctness.

File: qc_tests/streaks.py (hunk 15 above)
Comment (rjhd2) [reply qc_tests/streaks.py:560]: I've remembered why "full" - for "full reprocessing"

**Comment by rjhd2** [comment]
Environment updated under #162.  Starting to work through remaining comments and suggestions.

File: tests/test_qc_tests/test_streaks.py [hunk 24]
```diff
@@ -0,0 +1,372 @@
...
+    :returns: Station
+    
+    """
+    obs_var = common.example_test_variable(name, data)
+    station = common.example_test_station(obs_var, times)
+
+    return station
+
+
+def _make_simple_masked_data(stop: int, step: float) -> np.array:
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply tests/test_qc_tests/test_streaks.py:57]: Thanks - I'd not been fully specific here.

File: tests/test_qc_tests/test_streaks.py [hunk 25]
```diff
@@ -0,0 +1,372 @@
...
+    obs_var.data[50:70] = 5
+
+    return station
+
+@pytest.fixture()
+def station_for_rsc_logic() -> qc_utils.Station:
+    def _make_rsc_station() -> qc_utils.Station:
+        return _make_station(_make_simple_masked_data(100, 1), "temperature")
+    return _make_rsc_station()
+
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply tests/test_qc_tests/test_streaks.py:82]: My first fixture, I must have misunderstood the example I was following

File: qc_tests/streaks.py [hunk 26]
```diff
@@ -37,46 +45,118 @@ def plot_streak(times, obs_var, streak_start, streak_end):
...
+
+    return this_var
+
+
 #************************************************************************
-def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostics=False):
+def get_repeating_streak_threshold(obs_var: utils.Meteorological_Variable,
+                                   config_dict: dict,
+                                   wind_speed: Optional[utils.Meteorological_Variable] = None,
+                                   plots: bool = False,
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply qc_tests/streaks.py:119]: Aah - brilliant.  I wasn't happy with the extra import, but hadn't spotted on StackOv. that there was a newer alternative to use

File: qc_tests/streaks.py [hunk 27]
```diff
@@ -93,102 +173,369 @@ def get_repeating_string_threshold(obs_var, config_dict, plots=False, diagnostic
...
+
+    else:
+        # store high value so threshold never reached
+        try:
+            config_dict[f"STREAK-{this_var.name}"]["Excess"] = utils.MDI
+        except KeyError:
+            CD_straight = {"Excess" : utils.MDI}
+            config_dict[f"STREAK-{this_var.name}"] = CD_straight
+
+    return # get_excess_streak_threshold
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply qc_tests/streaks.py:248]: This is a hangover of how I used to code - coming from Fortran and not being exposed to PEP etc, I felt it was always useful to show that a function had ended/thought I needed to return.  Also, my functions in the past were over-long, so it was helpful if the start was off the top of the screen.  Clearly, that's not what is good style or practice, and I'm trying to refactor as part of these changes, but haven't necessarily found all of the errant `return`s

File: qc_tests/streaks.py [hunk 28]
```diff
@@ -203,29 +550,57 @@ def rsc(station, var_list, config_dict, full=False, plots=False, diagnostics=Fal
     for var in var_list:
 
         obs_var = getattr(station, var)
+        if obs_var.name == "wind_direction":
+            wind_speed = copy.deepcopy(getattr(station, "wind_speed"))
+        else:
+            wind_speed = None
 
         # need to have at least two observations to get a streak
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply qc_tests/streaks.py:558]: Thanks - I'd caught some of these in this set of edits and in other parallel PRs, but missed this one, and seemingly others too.

File: qc_tests/streaks.py [hunk 29]
```diff
@@ -203,29 +550,57 @@ def rsc(station, var_list, config_dict, full=False, plots=False, diagnostics=Fal
...
+            repeating_day(obs_var, station, config_dict, determine_threshold=False, plots=plots, diagnostics=diagnostics)
 
     return # rsc
 
 #************************************************************************
 if __name__ == "__main__":
 
-    print("checking repeated strings")
+    print("checking repeated streaks")
 #************************************************************************
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply qc_tests/streaks.py:606]: Correct - it's just to have something there.  You're right, it could be removed or at least the text improved to say what is happening (rather than implying that things are running when they're not!).  I might leave, and raise an issue (#163) so it's done consistently across all of these qc-check files.

File: tests/test_qc_tests/common.py [hunk 30]
```diff
@@ -53,3 +77,14 @@ def example_test_station(variable: utils.Meteorological_Variable,
     station.hours = np.array(times.dt.hour)
     
     return station
+
+
+def generate_streaky_data(data: np.array, starts_lengths: dict) -> np.array:
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply tests/test_qc_tests/common.py:82]: Thank you !

File: tests/test_qc_tests/test_streaks.py [hunk 31]
```diff
@@ -0,0 +1,372 @@
...
+    data = common.generate_streaky_data(data, EXCESS_STREAK_STARTS_LENGTHS)
+
+    # Add some years with streaks
+    for y in range(10, 15):
+        all_data = np.append(all_data, data)
+        years = np.append(years, (2000*np.ones(data.shape[0]))+y)
+        expected_proportions += [np.sum([v for _, v in EXCESS_STREAK_STARTS_LENGTHS.items()])/data.shape[0]]
+
+    station = _make_station(all_data, "temperature")
+    this_var = getattr(station, "temperature")
```
[the remark this answers is not in the document]
Comment (rjhd2) [reply tests/test_qc_tests/test_streaks.py:158]: Probably a good thing for me to do that to ensure that the test is working as I have intended!

**Review by rjhd2 — COMMENTED** [review]
Addressing most of @pwhybra's comments.  Will walk through and talk about the remaining in person, and see which other changes are suggested.

File: qc_tests/streaks.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (pwhybra) [reply qc_tests/streaks.py]: No problem, a very small nit by me!

File: tests/test_qc_tests/test_streaks.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (pwhybra) [reply tests/test_qc_tests/test_streaks.py]: There are other type hints I notice but unrelated to this PR changes and  I see you are fixing the type hints in another PR so will resolve here.

**Review by pwhybra — APPROVED** [review]
@rjhd2 has made a lot of requested changes, and other suggestions added as issues that will be changed in later PRs. Happy to get this one in to avoid it getting too bloated!
